contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,185 @@
+
+Several modifications to Design 1:
+* For discovery, do not rely on worker to report its identity as it can be compromised. Identify workers using the Endpoints object (pod name and node name).
+* In future versions, master goroutines should be event-driven, as constant polling is not scalable to large number of nodes. A worker pings the master when it's ready to report status, and master makes a separate call to fetch status from the worker. Worker must continuously ping master until it receives a status request to ensure master recognizes the status update. As a safety check, master should also poll workers periodically, but with very low frequency.
+  * It's not as secure to have workers push status directly because a worker can't be accurately identified from a gRPC request (the request IP can be spoofed).
+  * It's more convenient to use gRPC server streams, but it's hard to scale this to thousands of nodes as streams require TCP connections to be kept open.
+
+**TODO**: Use network policy to limit pod communication to only within the provisioner system?
+
+## Design 5: Worker as Jobs, Watcher in DaemonSet
+One of the major issues with previous approaches is worker identity - when a worker tries to communicate with master, it has to prove it's the worker it claims to be. To fix these issues, each worker must maintain different credentials.
+
+Alternatively, master can actively create worker Jobs to perform discovery and delete rather than having long-running worker pods in a DaemonSet sending data to the master. For discovery, workers can store the list of local volumes in its container STDOUT (**TODO**: decide on a form of structured output rather than plain text). For deletion, the status of the Job acts as an indication of deletion state. Master has perfect knowledge of which Job runs on which node, so if any worker is compromised, the damage is contained within the node the worker is scheduled on.
+
+To trigger discovery, we have local volume directory watcher pod running in a DaemonSet. It keeps an in-memory cache of local volumes and periodically checks the volume directory for changes. If there is a change, it notifies master through either gRPC or a CRD notification object, and master can then spawn a discovery Job.
+
+To ensure the provisioner system has up-to-date state on initialization, the following needs to be done:
+* When the master initializes, it spawns discovery jobs on every node to make sure it knows about all local volumes in the cluster.
+* When a watcher initializes, it should notify master so a discovery Job can be spawned to fetch the latest list of local volumes.
+
+Master can be made completely event driven using Job informers.
+
```

Comment:
Here's a 6th design, which I'm hesitant to suggest because it's a huge hack, but maybe more performant / reliable than 5:

Basically, handle master to worker communication via the exec/attach/portforward API. This interface is handled directly by the kubelet via the CRI (kubelet <-> pod communication over a unix domain socket), and the kubelet <-> master (apiserver) communication with mutual TLS. There would be bunch of proxy overhead though, so I don't think you would want to keep the stream open always. If performance was an issue the master provisioner could authenticate to the kubelet directly, but it would be much simpler to just communicate via the apiserver.